Unexpected "Invalid master password" error

Hello! Firstly, thanks for the efforts in building this workflow.

Unfortunately, it doesn't work for me when using the latest version of both the workflow (2.4.6) and Bitwarden CLI (2022.10.0). And the error "Invalid master password" is clearly hiding something more subtle since the master password is correct.

What I tried so far:

1. Logging with the password and API key (neither works)
2. Logging out from CLI with `bw logout` (from the CLI, I can log in and then unlock without problems)
3. Logging out/in from the workflow's auth settings
4. Deleting the workflow's cache and the `data.json` file located in `~/Library/Application Support`

The log in Alfred debugger isn't that useful since it reports only the "Invalid master password" error. Here it's, just in case:

```
[16:44:55.520] ERROR: Bitwarden v2[Run Script] 🍺
16:44:45 workflow.go:328: -------- Bitwarden v2/2.4.6 (AwGo/0.27.1) --------
16:44:45 main.go:162: &main.options{Search:false, Config:false, SetConfigs:false, Auth:false, OnOffConfigs:false, AuthConfig:false, Lock:false, Icons:false, Folder:false, Unlock:true, Login:false, Logout:false, Sync:false, Open:false, GetItem:false, Force:false, Totp:false, Last:false, Background:false, Id:"", Query:"", Attachment:"", Output:""}
16:44:45 main.go:164: args=[]string{"-unlock"} => []string{}
16:44:45 main.go:165: (main.config) {
 AutoFetchIconCacheAge: (int) 1440,
 AutoFetchIconMaxCacheAge: (time.Duration) 24h0m0s,
 BwconfKeyword: (string) (len=9) ".bwconfig",
 BwauthKeyword: (string) (len=7) ".bwauth",
 BwKeyword: (string) (len=3) ".bw",
 BwfKeyword: (string) (len=4) ".bwf",
 BwExec: (string) (len=2) "bw",
 BwDataPath: (string) "",
 Debug: (bool) false,
 Email: (string) (len=19) "xxx@yyyy.zz",
 EmailMaxWait: (int) 15,
 EmptyDetailResults: (bool) false,
 IconCacheAge: (int) 43200,
 IconCacheEnabled: (bool) true,
 IconMaxCacheAge: (time.Duration) 720h0m0s,
 MaxResults: (int) 1000,
 Mod1: (string) (len=3) "alt",
 Mod1Action: (string) (len=13) "username,code",
 Mod2: (string) (len=5) "shift",
 Mod2Action: (string) (len=3) "url",
 Mod3: (string) (len=4) "ctrl",
 Mod3Action: (string) (len=4) "totp",
 Mod4: (string) (len=7) "cmd,opt",
 Mod4Action: (string) (len=4) "more",
 Mod5: (string) (len=9) "cmd,shift",
 Mod5Action: (string) (len=5) "webui",
 NoModAction: (string) (len=13) "password,card",
 OpenLoginUrl: (bool) true,
 OutputFolder: (string) (len=25) "/Users/edoardo/Downloads/",
 Path: (string) (len=83) "/usr/bin:/usr/local/bin:/usr/local/sbin:/usr/local/share/npm/bin:/usr/bin:/usr/sbin",
 ReorderingDisabled: (bool) true,
 Server: (string) "",
 Sfa: (bool) true,
 SfaMode: (int) 0,
 SkipTypes: (string) "",
 TitleWithUser: (bool) true,
 TitleWithUrls: (bool) false,
 UseApikey: (bool) true,
 WebUiURL: (string) (len=27) "https://vault.bitwarden.com"
}
16:44:51 bitwarden.go:398: [DEBUG] ==> first few chars of the password is  xxx
16:44:55 utils.go:32: [ERROR] ==>  [Invalid master password.]
16:44:55 utils.go:44: [ERROR] ==> Exit code 1. Unlocking Bitwarden failed. Err: Invalid master password.
16:44:55 workflow.go:376: [ERROR] Unlocking Bitwarden failed. Error:
Invalid master password.
16:44:55 workflow.go:379: Get help at https://github.com/blacs30/bitwarden-alfred-workflow/issues
16:44:55 workflow.go:402: ------------------ 9.953131495s ------------------
[16:44:55.523] Bitwarden v2[Run Script] Processing complete
[16:44:55.530] Bitwarden v2[Run Script] Passing output 'Unlocking Bitwarden failed. Error:
Invalid master password.' to Post Notification
```

I'm not proficient in Go, hence it seemed a waste of time to dig into the source code and try to understand its inner working. But if there's any other info I can provide to help debugging, please let me know.

And thanks in advance for any hints!

--- a/src/bitwarden.go
+++ b/src/bitwarden.go
@@ -385,26 +385,22 @@ func runUnlock() {
 	)
 
 	// set the password from the returned slice
-	password := ""
-	if len(pw) > 0 {
-		password = pw
-	} else {
+	if len(pw) < 1 {
 		wf.Fatal("No Password returned.")
 	}
 
-	// remove newline characters
-	password = strings.TrimRight(password, "\r\n")
-	if wf.Debug() {
-		log.Println("[DEBUG] ==> first few chars of the password is ", password[0:2])
-	}
+	os.Setenv("PASS", pw)
 
 	// Unlock Bitwarden now
 	message := "Unlocking Bitwarden failed."
-	args := fmt.Sprintf("%s unlock --raw %s", conf.BwExec, password)
+	args := fmt.Sprintf("%s unlock --raw --passwordenv PASS", conf.BwExec)
+	debugLog(fmt.Sprintf("bw unlock command is %s", args))
 	tokenReturn, err := runCmd(args, message)
+	os.Unsetenv("PASS")
 	if err != nil {
 		wf.FatalError(err)
 	}
+
 	// set the password from the returned slice
 	token := ""
 	if len(tokenReturn) > 0 {
@@ -416,9 +412,7 @@ func runUnlock() {
 	if err != nil {
 		log.Println(err)
 	}
-	if wf.Debug() {
-		log.Println("[DEBUG] ==> first few chars of the token is ", token[0:2])
-	}
+	debugLog(fmt.Sprintf("first few chars of the token is %s", token[0:2]))
 
 	if conf.UseApikey {
 		// Writing the sync-cache because we have unlocked the vault in apikey mode
@@ -451,28 +445,21 @@ func runLogin() {
 		wf.Fatal("No email configured.")
 	}
 
-	// set the password from the returned slice
-	password := ""
-
 	if !conf.UseApikey {
 		_, pw, _ := zenity.Password(
 			zenity.Title(fmt.Sprintf("Login account %s", email)),
 		)
-		if len(pw) > 0 {
-			password = pw
-		} else {
+		if len(pw) < 1 {
 			return
 		}
-		password = strings.TrimRight(password, "\r\n")
-		if wf.Debug() {
-			log.Println("[DEBUG] ==> first few chars of the password is ", password[0:2])
-		}
+		os.Setenv("PASS", pw)
 	}
 
-	args := fmt.Sprintf("%s login %s %s", conf.BwExec, email, password)
+	args := fmt.Sprintf("%s login %s --passwordenv PASS", conf.BwExec, email)
+	debugLog(fmt.Sprintf("bw login command is %s", args))
 
-	log.Println("Use apikey", conf.UseApikey)
 	if conf.UseApikey {
+		log.Println("Use apikey", conf.UseApikey)
 		client_id, _ := zenity.Entry("Enter API Key client_id:",
 			zenity.Title(fmt.Sprintf("Login account %s", email)))
 		if len(client_id) < 1 {
@@ -505,7 +492,7 @@ func runLogin() {
 		} else if sfaMode == 1 {
 
 			emailMessage := "Failed to request Bitwarden email token."
-			emailArgs := fmt.Sprintf("%s login %s %s --raw --method %d", conf.BwExec, email, password, sfaMode)
+			emailArgs := fmt.Sprintf("%s login %s --raw --method %d --passwordenv PASS", conf.BwExec, email, sfaMode)
 			emailReturn, err := runCmdWithContext(conf.EmailMaxWait, emailArgs, emailMessage)
 			if err != nil {
 				wf.FatalError(err)
@@ -531,46 +518,44 @@ func runLogin() {
 			wf.Fatal("No 2FA code returned.")
 		}
 
-		args = fmt.Sprintf("%s login %s %s --raw --method %d --code %s", conf.BwExec, email, password, sfaMode, sfaCode)
+		args = fmt.Sprintf("%s login %s --passwordenv PASS --raw --method %d --code %s", conf.BwExec, email, sfaMode, sfaCode)
 	}
 
 	message := "Login to Bitwarden failed."
 	tokenReturn, err := runCmd(args, message)
+	os.Unsetenv("PASS")
 	if err != nil {
 		wf.FatalError(err)
 	}
 
 	// If we use the api key no token is returned, we first need to run unlock
-
 	if conf.UseApikey {
 		fmt.Println("APIKEY.")
 		return
-	} else {
-		// set the token from the returned result
-		token := ""
-		if len(tokenReturn) > 0 {
-			token = tokenReturn[0]
-		} else {
-			wf.Fatal("No token returned after unlocking.")
-		}
-		err = alfred.SetToken(wf, token)
-		if err != nil {
-			log.Println(err)
-		}
-		if wf.Debug() {
-			log.Println("[ERROR] ==> first few chars of the token is ", token[0:2])
-		}
-
-		// Writing the sync-cache because data is synced for Yubikey and Authenticator login
-		// Just the APIKEY login needs a separate unlock and therefore sync
-		err = wf.Cache.Store(SYNC_CACHE_NAME, []byte("sync-cache"))
-		if err != nil {
-			log.Println(err)
-		}
-
-		// Creating the items cache
-		runCache()
 	}
+
+	// set the token from the returned result
+	token := ""
+	if len(tokenReturn) > 0 {
+		token = tokenReturn[0]
+	} else {
+		wf.Fatal("No token returned after unlocking.")
+	}
+	err = alfred.SetToken(wf, token)
+	if err != nil {
+		log.Println(err)
+	}
+	debugLog(fmt.Sprintf("first few chars of the token is %s", token[0:2]))
+
+	// Writing the sync-cache because data is synced for Yubikey and Authenticator login
+	// Just the APIKEY login needs a separate unlock and therefore sync
+	err = wf.Cache.Store(SYNC_CACHE_NAME, []byte("sync-cache"))
+	if err != nil {
+		log.Println(err)
+	}
+
+	// Creating the items cache
+	runCache()
 
 	searchAlfred(conf.BwKeyword)
 	fmt.Println("Logged In.")

--- a/workflow/info.plist
+++ b/workflow/info.plist
@@ -2234,7 +2234,7 @@ Caching of the secret/item names (not the secret values itself) are cached so th
 		<string>WEBUI_URL</string>
 	</array>
 	<key>version</key>
-	<string>2.4.6</string>
+	<string>2.4.7</string>
 	<key>webaddress</key>
 	<string>https://github.com/blacs30/bitwarden-alfred-workflow</string>
 </dict>

